src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -203,15 +199,79 @@ void NnetDecodableBase::GetCurrentIvector(int32 output_t_start,
     }
     ivector_frame = online_ivector_feats_->NumRows() - 1;
+  } else if (ivector_frame < 0) {
+    ivector_frame = 0;
   }
-  *ivector = online_ivector_feats_->Row(ivector_frame);
+  ivector->CopyFromVec(online_ivector_feats_->Row(ivector_frame));
 }
 
+void NnetDecodableBase::GetIvectorsForFrames(int32 output_t_start,
+                                             int32 num_output_frames,
+                                             Matrix<BaseFloat> *ivectors) {
+  // if no ivectors have been specified either as online ivectors or as ivector
+  // for utterance, just return and leave the size of ivectors matrix as 0 
+  if (ivector_ == NULL && online_ivector_feats_ == NULL) {
+    KALDI_ASSERT(ivectors->NumCols() == 0);
+    return;
+  }
+  const int32 left_context = nnet_left_context_ + opts_.extra_left_context,
+        right_context = nnet_right_context_ + opts_.extra_right_context;
+  // frame_to_search is the frame that we want to get the most recent iVector
+  // for. In single ivector case, We choose a point near the middle of the
+  // current window, the concept being that this is the fairest comparison to
+  // nnet2. In multiple ivectors case, we just choose frames at whole multiples
+  // of ivector_period_. Obviously we could do better by always taking the
+  // last frame's iVector, but decoding with 'online' ivectors is only really
+  // a mechanism to simulate online operation.
+  if (ivector_period_ == 0) { // single ivector case
+    ivectors->Resize(1, online_ivector_feats_->NumCols());
+    int32 frame_to_search = output_t_start + num_output_frames / 2;
+    SubVector<BaseFloat> sub_ivector = ivectors->Row(0);
+    GetCurrentIvector(frame_to_search, &sub_ivector);
+  } else { // multiple ivectors case
+    // num_ivectors_in_left_context is the num of ivectors for frames
+    // whose "t" index < 0. It is used to compute the initial value of
+    // frame_to_search, which is the frame that we get the left-most ivector for
+    // num_ivectors is the num of ivectors for the entire chunk.
+    // Both of them are computed according to how Round descriptor works, which
+    // basically returns floor(t / <t-modulus>) * <t-modulus>.
+    // (The assumption is that the t index of the first outptut frame is 0)
+    int32 num_ivectors_in_left_context =
+        std::ceil(1.0 * left_context / ivector_period_);
+    int32 num_ivectors = num_ivectors_in_left_context +
+        (num_output_frames + right_context - 1) / ivector_period_ + 1;
+    ivectors->Resize(num_ivectors, online_ivector_feats_->NumCols());
+    int32 frame_to_search = output_t_start -
+                            num_ivectors_in_left_context * ivector_period_;
+    for (int32 n = 0; n < num_ivectors; n++) {
+      SubVector<BaseFloat> sub_ivector = ivectors->Row(n);
+      GetCurrentIvector(frame_to_search, &sub_ivector);
+      frame_to_search += ivector_period_;
+    }
+  }
+}
+
+void NnetDecodableBase::GenerateIndexesForIvectors(
```

Comment:
OK, I see now that doing it together is impractical, because you call GenerateIndexesForIvectors() while doing the compilation.  But it would be better to get directly from the ComputationRequest, the frames on which you need the iVectors.  That way, there is no need to ensure that two separate functions remain exactly in sync.